pineappl/src/grid.rs

```diff
@@ -1815,6 +1779,288 @@ impl Grid {
+                                .find_map(|((&(pa0, pa1), opa), (&(pb0, pb1), opb))| {
+                                    (pa0 == pida0 && pa1 == pida1 && pb0 == pidb0 && pb1 == pidb1)
+                                        .then(|| (opa, opb))
+                                })
+                                .map(|(opa, opb)| (fk_table, opa, opb))
+                        },
+                    ) {
+                        let mut result = Array2::zeros((info.x0.len(), info.x0.len()));
+
+                        for imu2 in 0..array.dim().0 {
+                            let opa = opa.index_axis(Axis(0), imu2);
+                            let opb = opb.index_axis(Axis(0), imu2);
+                            let arr = array.index_axis(Axis(0), imu2);
+
+                            result += &opa.dot(&arr.dot(&opb.t()));
+                        }
+
+                        fk_table.scaled_add(factor, &result);
+                    }
+                }
+            }
+
+            sub_fk_tables.extend(tables.into_iter().map(|table| {
+                ImportOnlySubgridV2::new(
+                    SparseArray3::from_ndarray(&table.insert_axis(Axis(0)), 0, 1),
+                    vec![Mu2 {
+                        // TODO: FK tables don't depend on the renormalization scale
+                        //ren: -1.0,
+                        ren: info.fac0,
+                        fac: info.fac0,
+                    }],
+                    info.x0.clone(),
+                    info.x0.clone(),
+                )
+                .into()
+            }));
+        }
+
+        let mut grid = Self {
+            subgrids: Array1::from_iter(sub_fk_tables.into_iter())
+                .into_shape((1, self.bin_info().bins(), lumi0.len()))
+                .unwrap(),
+            lumi: lumi0.iter().map(|&(a, b)| lumi_entry![a, b, 1.0]).collect(),
+            bin_limits: self.bin_limits.clone(),
+            orders: vec![Order::new(0, 0, 0, 0)],
+            subgrid_params: SubgridParams::default(),
+            more_members: self.more_members.clone(),
+        };
+
+        // write additional metadata
+        grid.set_key_value("lumi_id_types", &info.lumi_id_types);
+
+        Ok(FkTable::try_from(grid).unwrap())
+    }
+
+    /// Converts this `Grid` into an [`FkTable`] using an evolution kernel operator (EKO) given as
+    /// `operator`. The dimensions and properties of this operator must be described using `info`.
+    /// The parameter `order_mask` can be used to include or exclude orders from this operation,
+    /// and must correspond to the ordering given by [`Grid::orders`]. Orders that are not given
+    /// are enabled, and in particular if `order_mask` is empty all orders are activated.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`GridError::EvolutionFailure`] if either the `operator` or its `info` is
+    /// incompatible with this `Grid`.
+    pub fn evolve(
```

Comment:
- This one
- `evolve_with_one`
- `evolve_with_two`

I would move into the `evolution.rs` module, into a second `impl Grid` block.
So, I would also revert the dependency: instead of using `evolution.rs` in here, I would use the `grid.rs` in the evolution module (such that the grid module is limited to the basic grid features).

Reply:
Of course tests should be moved accordingly

Reply:
This is good idea! I didn't think of that myself before because you must have the `impl` block of a `struct` in the same module as the `struct` is defined. However, we can use a free-standing function like all the other members of `mod evolution`.

Reply:
> you must have the `impl` block of a `struct` in the same module as the `struct` is defined.

I'm not aware of this as a good practice (in case do you know the reason? or a reference?), but I can tell you that is definitely not enforced by the language. 

I have a working example, but unfortunately in a private repo.

Reply:
You're right: <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=594782c57f2c7dae572c17faa4e4e734>. However, you can't access private members. In commit 8d5fdca8c66b5c2afeb24a09141d17752c78041e I used free functions which are just as good.

Reply:
Ah yes, you're right.

Your solution is perfect. Later on we can decide to move them in an `impl` block, but is reasonable that they will rely on `pub(crate)` attributes and methods (this way we limit the interdependency of elements).

Reply:
I think in terms of encapsulation it's even better to have a free-standing function than a method.